src/runtime_src/core/include/experimental/xrt_queue.h

```diff
@@ -0,0 +1,297 @@
+#ifdef __cplusplus
+# include <algorithm>
+# include <future>
+# include <memory>
+#endif
+
+#ifdef __cplusplus
+
+namespace xrt {
+
+/**
+ * class queue -- producer / consumer queue for tasks
+ *
+ * Used for sequencing operations in order of enqueuing.
+ *
+ * A queue has exactly one consumer which is a separate thread created
+ * when the queue is constructed.
+ *
+ * When an opeation is enqueued on the queue an event is returned to
+ * the caller.  This event can be enqueued in a different queue, which
+ * will then wait for the former to complete the operaiton associated
+ * with the event.
+ */
+class queue_impl;
+class queue
+{
+  friend class queue_impl;
+
+  // class task - wraps a typed callable operation
+  //
+  // A task wraps a callers typed operation such that it
+  // can be inserted into a queue.
+  //
+  // Tasks should be synchronous operations.
+  class task
+  {
+    struct task_iholder
+    {
+      virtual ~task_iholder() {};
+      virtual void execute() = 0;
+    };
+
+    // Wrap typed operation
+    template <typename Callable>
+    struct task_holder : public task_iholder
+    {
+      Callable m_held;
+
+      task_holder(Callable&& t)
+        : m_held(std::move(t))
+      {}
+
+      void execute()
+      {
+        m_held();
+      }
+    };
+
+    std::unique_ptr<task_iholder> m_content;
+
+  public:
+    task() : m_content(nullptr)
+    {}
+
+    task(task&& rhs) : m_content(std::move(rhs.m_content))
+    {}
+
+    // task() - task constructor for synchronous operation
+    //
+    // @c : callable object, a std::packaged_task
+    template <typename Callable>
+    task(Callable&& c)
+      : m_content(new task_holder<Callable>(std::forward<Callable>(c)))
+    {}
+
+    task&
+    operator=(task&& rhs)
+    {
+      m_content = std::move(rhs.m_content);
+      return *this;
+    }
+
+    operator bool() const
+    {
+      return m_content!=nullptr;
+    }
+
+    void
+    execute()
+    {
+      m_content->execute();
+    }
+  };  // class queue::task
+
+public:
+
+  /**
+   * class event - type erased std::shared_future
+   *
+   * Wraps typed future return value returned when
+   * enqueueing an operation.
+   *
+   * Returned futures implicitly convert to an event and as such
+   * enqueuue return values can be stored in an event container if
+   * necessary.
+   *
+   * The event object is not needed where the typed future can be
+   * used.
+   */
+  class event
+  {
+    struct event_iholder
+    {
+      virtual ~event_iholder() {};
+      virtual void wait() const = 0;
+    };
+
+    // Wrap tyoed future
+    template <typename ValueType>
+    struct event_holder : public event_iholder
+    {
+      std::shared_future<ValueType> m_held;
+
+      event_holder(std::shared_future<ValueType>&& e)
+        : m_held(std::move(e))
+      {}
+
+      void wait() const
+      {
+        m_held.wait();
+      }
+    };
+
+    std::shared_ptr<event_iholder> m_content;
+
+  public:
+    event() : m_content(nullptr)
```

Comment:
```suggestion
    event() : m_content{}
```
or you can even just use the default constructor